File: common/go/discovery/registry_impl.go
package discovery

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/gmaker/luffa/common/go/logger"
	"github.com/gmaker/luffa/common/go/registry"
	pb "github.com/gmaker/luffa/gen/go/registry"
)

// RegistryImpl 基于自研 Registry 的 ServiceDiscovery 实现
type RegistryImpl struct {
	client        *registry.Client
	heartbeatStop chan struct{}
	heartbeatWG   sync.WaitGroup
	nodeID        string
}

// NewRegistryImpl 创建 Registry 模式的服务发现实例
func NewRegistryImpl(addrs []string) *RegistryImpl {
	return &RegistryImpl{
		client:        registry.NewClient(addrs),
		heartbeatStop: make(chan struct{}),
	}
}

// Register 注册节点并启动后台心跳保活
func (r *RegistryImpl) Register(ctx context.Context, node NodeInfo) error {
	if err := r.client.Connect(); err != nil {
		return fmt.Errorf("connect registry failed: %w", err)
	}

	pbNode := &pb.NodeInfo{
		ServiceType: node.ServiceType,
		NodeId:      node.NodeID,
		Host:        node.Host,
		Port:        node.Port,
		Metadata:    node.Metadata,
		LoadScore:   node.LoadScore,
		RegisterAt:  node.RegisterAt,
	}

	r.nodeID = node.NodeID
	if _, err := r.client.RegisterWithRetry(pbNode, 5); err != nil {
		return fmt.Errorf("register failed: %w", err)
	}

	// 启动心跳 goroutine
	r.heartbeatWG.Add(1)
	go func() {
		defer r.heartbeatWG.Done()
		ticker := time.NewTicker(5 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				_, err := r.client.Heartbeat(r.nodeID)
				if err != nil {
					logger.Warnf("registry heartbeat failed: %v", err)
				}
			case <-r.heartbeatStop:
				return
			}
		}
	}()

	return nil
}

// Deregister 注销节点并停止心跳
func (r *RegistryImpl) Deregister(ctx context.Context, nodeID string) error {
	close(r.heartbeatStop)
	r.heartbeatWG.Wait()
	// 自研 Registry 目前无显式注销接口，依赖心跳超时自动清理
	return nil
}

// Discover 发现服务节点
func (r *RegistryImpl) Discover(ctx context.Context, serviceType string) ([]NodeInfo, error) {
	list, err := r.client.Discover(serviceType)
	if err != nil {
		return nil, err
	}

	var nodes []NodeInfo
	for _, n := range list.Nodes {
		nodes = append(nodes, pbNodeToNodeInfo(n))
	}
	return nodes, nil
}

// Watch 监听节点变更
func (r *RegistryImpl) Watch(ctx context.Context, serviceTypes []string, callback func(NodeEvent)) error {
	_, err := r.client.Subscribe(serviceTypes, func(ev *pb.NodeEvent) {
		callback(pbEventToNodeEvent(ev))
	})
	return err
}

// Close 关闭 Registry 连接
func (r *RegistryImpl) Close() error {
	select {
	case <-r.heartbeatStop:
		// 已关闭
	default:
		close(r.heartbeatStop)
	}
	r.heartbeatWG.Wait()
	r.client.Close()
	return nil
}

func pbNodeToNodeInfo(n *pb.NodeInfo) NodeInfo {
	return NodeInfo{
		ServiceType: n.ServiceType,
		NodeID:      n.NodeId,
		Host:        n.Host,
		Port:        n.Port,
		Metadata:    n.Metadata,
		LoadScore:   n.LoadScore,
		RegisterAt:  n.RegisterAt,
	}
}

func pbEventToNodeEvent(ev *pb.NodeEvent) NodeEvent {
	var t NodeEventType
	switch ev.Type {
	case pb.NodeEvent_JOIN:
		t = NodeEventJoin
	case pb.NodeEvent_LEAVE:
		t = NodeEventLeave
	case pb.NodeEvent_UPDATE:
		t = NodeEventUpdate
	}
	return NodeEvent{
		Type: t,
		Node: pbNodeToNodeInfo(ev.Node),
	}
}

File: common/cpp/discovery/registry_impl.cpp
#include "registry_impl.hpp"
#include "registry.pb.h"
#include <iostream>

namespace gs {
namespace discovery {

RegistryImpl::RegistryImpl(net::async::AsyncEventLoop* loop,
                           const std::vector<std::pair<std::string, uint16_t>>& addrs)
    : client_(std::make_unique<gs::registry::RegistryClient>(loop, addrs)) {}

RegistryImpl::~RegistryImpl() {
    Close();
}

bool RegistryImpl::Register(const NodeInfo& node) {
    if (!client_->Connect()) {
        std::cerr << "RegistryImpl::Register: connect failed" << std::endl;
        return false;
    }

    PbNodeInfo pb;
    pb.set_service_type(node.service_type);
    pb.set_node_id(node.node_id);
    pb.set_host(node.host);
    pb.set_port(node.port);
    for (const auto& [k, v] : node.metadata) {
        (*pb.mutable_metadata())[k] = v;
    }
    pb.set_load_score(node.load_score);
    pb.set_register_at(node.register_at);

    ::registry::Result res;
    int maxRetries = 5;
    bool ok = false;
    for (int i = 0; i < maxRetries; ++i) {
        if (client_->Register(pb, &res)) {
            ok = true;
            break;
        }
        if (i < maxRetries - 1) {
            std::this_thread::sleep_for(std::chrono::seconds(i + 1));
        }
    }
    if (!ok) {
        std::cerr << "RegistryImpl::Register: register failed after retries" << std::endl;
        return false;
    }

    // 启动心跳线程
    heartbeat_stop_ = false;
    heartbeat_thread_ = std::thread([this, node_id = node.node_id]() {
        HeartbeatLoop(node_id);
    });

    return true;
}

bool RegistryImpl::Deregister(const std::string& node_id) {
    (void)node_id;
    // 自研 Registry 目前无显式注销接口，依赖心跳超时自动清理
    heartbeat_stop_ = true;
    if (heartbeat_thread_.joinable()) {
        heartbeat_thread_.join();
    }
    return true;
}

bool RegistryImpl::Discover(const std::string& service_type, std::vector<NodeInfo>& out) {
    ::registry::NodeList list;
    if (!client_->Discover(service_type, &list)) {
        return false;
    }
    for (int i = 0; i < list.nodes_size(); ++i) {
        out.push_back(PbNodeToNodeInfo(list.nodes(i)));
    }
    return true;
}

bool RegistryImpl::Watch(const std::vector<std::string>& service_types, EventCallback callback) {
    ::registry::SubscribeRes res;
    bool ok = client_->Subscribe(service_types, &res,
        [this, callback](const PbNodeEvent& ev) {
            callback(PbEventToNodeEvent(ev));
        });
    return ok;
}

void RegistryImpl::Close() {
    heartbeat_stop_ = true;
    if (heartbeat_thread_.joinable()) {
        heartbeat_thread_.join();
    }
    client_->Close();
}

void RegistryImpl::HeartbeatLoop(const std::string& node_id) {
    while (!heartbeat_stop_) {
        ::registry::Result res;
        if (!client_->Heartbeat(node_id, &res)) {
            std::cerr << "RegistryImpl::HeartbeatLoop: heartbeat failed" << std::endl;
        }
        for (int i = 0; i < 50 && !heartbeat_stop_; ++i) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
    }
}

NodeInfo RegistryImpl::PbNodeToNodeInfo(const PbNodeInfo& pb) {
    NodeInfo node;
    node.service_type = pb.service_type();
    node.node_id = pb.node_id();
    node.host = pb.host();
    node.port = pb.port();
    for (const auto& [k, v] : pb.metadata()) {
        node.metadata[k] = v;
    }
    node.load_score = pb.load_score();
    node.register_at = pb.register_at();
    return node;
}

NodeEvent RegistryImpl::PbEventToNodeEvent(const PbNodeEvent& pb) {
    NodeEvent ev;
    switch (pb.type()) {
        case PbNodeEvent::JOIN:   ev.type = NodeEventType::Join;   break;
        case PbNodeEvent::LEAVE:  ev.type = NodeEventType::Leave;  break;
        case PbNodeEvent::UPDATE: ev.type = NodeEventType::Update; break;
        default:                  ev.type = NodeEventType::Update; break;
    }
    ev.node = PbNodeToNodeInfo(pb.node());
    return ev;
}

} // namespace discovery
} // namespace gs
